### ISSUE TITLE
Git file history is lost for some file 

### ISSUE THREAD

**Issue by kartikag01** [gregjacobs/js-to-ts-converter#19]
hey, when i run this script on about 1000 Js files, git tracked lot many file as rename, but for about 150 files git says `.js` files are deleted and created new  `.ts` files

Is there any way we can preserve git history ?

**Comment by gregjacobs** [comment]
Hey Kartik,

Git does rename detection based on file content similarity, so if the file has changed drastically, git may consider the file a deletion+addition. However, the threshold that git uses to determine if a file is "similar enough" can be changed at any time, even after you commit. 

The commit object itself for the renaming of files doesn't actually store any information about renames; it only stores the deletes/adds. It's only when you're viewing the history of a file does git actually compare the file deletions and additions in any given commit by file contents to determine if two files are the same file that has actually been renamed. So this is both good and bad news: good in that you can actually tweak the threshold of what git should consider two files "similar enough" that they'll be considered a rename when running commands like 'git diff' (git diff --find-renames=40% would mean files need to be at least 40% similar to be considered a rename), but bad in that other team members will need to have the exact same rename threshold setting to find those renames later in time.

Bottom line though here is: it doesn't actually matter what git sees as a rename or not when you commit those 1000 files. You'll need to configure your tools (the git CLI, any gui git tool you use) to be have a more relaxed rename threshold when you search through your file history later. I believe git defaults to 50% similarity of file contents in order to consider a file a rename, so you'll need to reduce that threshold later.

Best,
Greg
